Rethrow BatchParserExceptions, swallow the stack trace and add the name of the file in error

jeffrosenberg opened the pull request:
Fixes #128

jeffrosenberg commented:
This is my attempt to throw a more user-friendly error when we get a BatchParserException, suggestions on how to improve it are welcome 😃

jmezach commented on src/DacpacTool/ScriptParser.cs:
```diff
@@ -24,8 +25,20 @@ private void Parse()
+                    _parser.Parse();
+                    _parsed = true;
+                }
+                catch (Microsoft.SqlTools.ServiceLayer.BatchParser.BatchParserException ex)
```
Do we need to fully qualify the exception type? There's already a using for this namespace in this file?

jmezach commented on src/DacpacTool/ScriptParser.cs:
```diff
@@ -24,8 +25,20 @@ private void Parse()
+                catch (Microsoft.SqlTools.ServiceLayer.BatchParser.BatchParserException ex)
+                {
+                    string mostRecentFile = _includedFileNames.LastOrDefault();
+                    if (mostRecentFile != null)
```
I would prefer not to have two different throw statements, but put the exception message into a variable (possibly using a ternary operator) and then a single throw statement.

jeffrosenberg replied:
heh, that was my original approach but I decided I didn't like it. 😄 I don't have a strong opinion, though, just thought it looked a little messy with the ternary. I'll go back to this approach with the next commit.

jmezach commented on test/DacpacTool.Tests/ScriptParserTests.cs:
```diff
@@ -162,5 +162,16 @@ public void CanGenerateScriptWithNestedIncludesContainingRelativePaths()
+
+            // Act / Assert
+            Should.Throw<System.InvalidOperationException>(() => collector.CollectFileNames())
+                .Message.ShouldBe("Incorrect syntax was encountered while parsing '\n'. File: DataScript.sql");
```
The message here feels a bit weird to me, especially with the `\n` before the dot. Is that deliberate?

jeffrosenberg replied:
The first half of the error comes from the test scenario I imported, per my previous comment, I just matched exactly what was coming back from the parser error.

jmezach commented on test/TestProjectWithIncludedSyntaxErrors/Post-Deployment/DataScript.sql:
```diff
@@ -0,0 +1,3 @@
+print 'Loading file '
```
Do we need all these test files just to trigger the error? Feels like a contrived example scenario that could be simplified for easier maintainability.

jmezach commented on test/TestProjectWithIncludedSyntaxErrors/TestProjectWithIncludedSyntaxErrors.csproj:
```diff
@@ -0,0 +1,44 @@
+<Project Sdk="MSBuild.Sdk.SqlProj/1.16.0-beta.4">
```
Perhaps we could setup this project to use the `Sdk.props` and `Sdk.targets` files directly as the other test projects do?

jmezach reviewed (changes requested):
@jeffrosenberg Thanks for this! Have a couple of comments.

jeffrosenberg commented:
@jmezach I've addressed most of your comments, except for the two related to the new test project; I didn't have time to work on that tonight. Please take a look at my comments and if you'd like me to give that another shot, I can definitely do that, but it's a holiday weekend here in the US so it might be a few days before I get that pushed. 🙂

jmezach commented:
Changes look good to me. I do think we should try to create the simplest test scenario that we possibly can, to avoid shooting ourselves in the foot later on, but obviously there's no need to rush this :). Enjoy the holiday weekend! I'll probably be a little bit less responsive in the coming weeks as well, since I got the keys to my new house last Thursday so there's a lot of things that will need to happen now.

jeffrosenberg commented:
Just checking in on this to let you know that I'll have to finish it up next week. I didn't get to this over the last couple of days and now I'm heading out for a few days of vacation. Not much left here, I'll definitely wrap it up next week 🙂 

> I'll probably be a little bit less responsive in the coming weeks as well, since I got the keys to my new house last Thursday

Congrats, @jmezach, enjoy!

jeffrosenberg commented:
@jmezach I've simplified the test scenarios. In doing so, I've realized that the scenario provided in the initial issue -- an empty include directive -- may be about all that's actually handled by this change. Adding an include to a non-existent script (properly) throws a `FileNotFoundException`. Adding an include to a script with syntax errors in it ignores the syntax errors entirely, I guess because all the parser is doing is traversing files and concatenating them.

I would love to be able to do more than this, particularly identifying syntax errors in pre- and post-deployment scripts, but that will require significantly more work and I'm not certain whether it's doable or not. My recommendation would be to merge this as it resolves the specific issue, and then I'd be open to adding a new issue to see if we can improve upon this further.

jmezach reviewed (approved):
@jeffrosenberg This looks good to me. I agree that it would be useful to have meaningful parser errors for included scripts but as you said it might be tricky to pull off. Perhaps we can do something with the code we're including from Microsoft, but we can always figure that out later. For now, let's get this merged and released :).
